From 895fcbd89333ac1d886b42de39d351aae2090acc Mon Sep 17 00:00:00 2001
From: Hannes Frederic Sowa <hannes@redhat.com>
Date: Thu, 18 Dec 2014 23:30:32 -0500
Subject: [scsi] do not issue SCSI RSOC command to Promise Vtrak E610f

Message-id: <b560705cde7c292e84895c87e3b2311cfb7b6b4c.1418941128.git.hannes@redhat.com>
Patchwork-id: 102323
O-Subject: [RHEL7.1 scsi 1/2] scsi: do not issue SCSI RSOC command to Promise Vtrak E610f
Bugzilla: 1175862
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>

Upstream Commit:
commit 0213436a2cc5e4a5ca2fabfaa4d3877097f3b13f
Author: Janusz Dziemidowicz <rraptorr@nails.eu.org>
Date:   Thu Jul 24 15:48:46 2014 +0200

    scsi: do not issue SCSI RSOC command to Promise Vtrak E610f

    Some devices don't like REPORT SUPPORTED OPERATION CODES and will
    simply timeout causing sd_mod init to take a very very long time.
    Introduce BLIST_NO_RSOC scsi scan flag, that stops RSOC from being
    issued. Add it to Promise Vtrak E610f entry in scsi scan
    blacklist. Fixes bug #79901 reported at
    https://bugzilla.kernel.org/show_bug.cgi?id=79901

    Fixes: 98dcc2946adb ("SCSI: sd: Update WRITE SAME heuristics")

    Signed-off-by: Janusz Dziemidowicz <rraptorr@nails.eu.org>
    Reviewed-by: Martin K. Petersen <martin.petersen@oracle.com>
    Cc: stable@vger.kernel.org
    Signed-off-by: Christoph Hellwig <hch@lst.de>

Signed-off-by: Hannes Frederic Sowa <hannes@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/scsi_devinfo.c b/drivers/scsi/scsi_devinfo.c
index f969aca..49014a1 100644
--- a/drivers/scsi/scsi_devinfo.c
+++ b/drivers/scsi/scsi_devinfo.c
@@ -222,6 +222,7 @@ static struct {
  {"PIONEER", "CD-ROM DRM-602X", NULL, BLIST_FORCELUN | BLIST_SINGLELUN},
  {"PIONEER", "CD-ROM DRM-604X", NULL, BLIST_FORCELUN | BLIST_SINGLELUN},
  {"PIONEER", "CD-ROM DRM-624X", NULL, BLIST_FORCELUN | BLIST_SINGLELUN},
+ {"Promise", "VTrak E610f", NULL, BLIST_SPARSELUN | BLIST_NO_RSOC},
  {"Promise", "", NULL, BLIST_SPARSELUN},
  {"QUANTUM", "XP34301", "1071", BLIST_NOTQ},
  {"REGAL", "CDC-4X", NULL, BLIST_MAX5LUN | BLIST_SINGLELUN},
diff --git a/drivers/scsi/scsi_scan.c b/drivers/scsi/scsi_scan.c
index 66b759a..f118c4e 100644
--- a/drivers/scsi/scsi_scan.c
+++ b/drivers/scsi/scsi_scan.c
@@ -901,6 +901,12 @@ static int scsi_add_lun(struct scsi_device *sdev, unsigned char *inq_result,
  if (*bflags & BLIST_USE_10_BYTE_MS)
   sdev->use_10_for_ms = 1;
 
+ /* some devices don't like REPORT SUPPORTED OPERATION CODES
+  * and will simply timeout causing sd_mod init to take a very
+  * very long time */
+ if (*bflags & BLIST_NO_RSOC)
+  sdev->no_report_opcodes = 1;
+
  /* set the device running here so that slave configure
   * may do I/O */
  ret = scsi_device_set_state(sdev, SDEV_RUNNING);
diff --git a/include/scsi/scsi_devinfo.h b/include/scsi/scsi_devinfo.h
index 447d2d7..7f7ec7f 100644
--- a/include/scsi/scsi_devinfo.h
+++ b/include/scsi/scsi_devinfo.h
@@ -32,4 +32,7 @@
 #define BLIST_ATTACH_PQ3 0x1000000 /* Scan: Attach to PQ3 devices */
 #define BLIST_NO_DIF  0x2000000 /* Disable T10 PI (DIF) */
 #define BLIST_SKIP_VPD_PAGES 0x4000000 /* Ignore SBC-3 VPD pages */
+
+#define BLIST_NO_RSOC  0x20000000 /* don't try to issue RSOC */
+
 #endif
-- 
1.7.1